### PR TITLE
[CDAP-1433] Consolidate Time Partitioned and Partitioned FileSets

### PR DESCRIPTION
**Pull request by anew**
- make TPFS a subclass of PFS
- add backward compatibility
- add unit tests
- remove unneeded code from explore
- change all getUnderlyingFileSet() to getEmbeddedFileSet()

### CODE REVIEW

**Comment by anew** [comment]
Build succeeded here: http://builds.cask.co/browse/CDAP-DUT762-3

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java [hunk 1]
```diff
@@ -434,15 +449,7 @@ static PartitionKey parseRowKey(byte[] rowKey, Partitioning partitioning) {
...
-    PartitionKey key;
-    try {
-      key = parseRowKey(row.getRow(), partitioning);
-    } catch (Exception e) {
-      LOG.debug(String.format("Failed to parse row key for partitioned file set '%s': %s",
-                              getName(), Bytes.toStringBinary(row.getRow())));
-      return false;
-    }
-    return filter.match(key);
+  protected Table getPartitionsTable() {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java]: the field is already protected, seems like we can remove this

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java [hunk 2]
```diff
@@ -102,7 +103,10 @@ public void addPartition(PartitionKey key, String path) {
               Bytes.toBytes(entry.getValue().toString()));            // "<string rep. of value>"
     }
     partitions.put(put);
+    addPartitionToExplore(key, path);
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java]: what happens if this fails? does the put to partitions get invalidated?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java [hunk 3]
```diff
@@ -187,21 +189,33 @@ public String getPartition(PartitionKey key) {
         if (row == null) {
           break;
         }
-        if (filter != null && !validateFilter(filter, row)) {
+        PartitionKey key;
+        try {
+          key = parseRowKey(row.getRow(), partitioning);
+        } catch (IllegalArgumentException e) {
+          if (!ignoreInvalidRowsSilently) {
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java]: so ignore silently just means we don't debug log but still ignore the row?  Is it expected that we won't be able to parse the row key?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java [hunk 4]
```diff
@@ -150,36 +157,31 @@ public String getPartition(PartitionKey key) {
...
+      @Override
+      public void consume(PartitionKey key, String path) {
+        paths.add(path);
       }
-      return paths;
-    } finally {
-      scanner.close();
-    }
+    });
+    return paths;
```
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java]: seems simpler to just do ```getPartitions(null).values()```

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java (hunk 3 above)
Comment (albertshau) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java]: ok i see it's the older data

**Comment by albertshau** [comment]
minor comment and one question, otherwise lgtm

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java (hunk 3 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java]: the old data format can collide with scan range of the new format. If compatibility is turned on, then this is expected and we ignore it. If it is not turned on, then we expect no legacy rows, we still ignore these rows but we it is worth logging them.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java (hunk 2 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java]: No. This is a common problem. We have a Jira open to make DDL operations transactional. Itis complex to do that and beyond the scope of this Jira. But I can add a TODO here.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java (hunk 1 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java]: ok

File: cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java (hunk 2 above)
Comment (anew) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java]: Jira is CDAP-1393

**Comment by albertshau** [comment]
:+1:

**Comment by anew** [comment]
thanks @albertshau - I will merge after the build succeeds

**Comment by anew** [comment]
Build succeeded here: http://builds.cask.co/browse/CDAP-DUT762
Merging.

**Pull request merged**